engines/director/lingo/lingo-bytecode.cpp

```diff
@@ -0,0 +1,311 @@
+						warning("Constant float start offset is out of bounds");
+						break;
+					} else if (value + 4 > constsStoreSize) {
+						warning("Constant float end offset is out of bounds");
+						break;
+					}
+					constant.u.f = *(float *)(constsStore + value);
+				}
+				break;
+			default:
+				warning("Unknown constant type %d", type);
+				break;
+		}
+
+		constsData.push_back(constant);
+	}
+	free(constsStore);
+
+	// parse each function!
+	// these are stored as a code storage area, followed by a reference table of 42 byte entries.
+
+	// copy the storage area first.
+	uint32 codeStoreSize = functionsOffset - codeStoreOffset;
+	stream.seek(codeStoreOffset);
+	byte *codeStore = (byte *)malloc(codeStoreSize);
+	stream.read(codeStore, codeStoreSize);
+
+	// read each entry in the function table.
+	stream.seek(functionsOffset);
+	for (uint16 i=0; i<functionsCount; i++) {
+		_currentScriptFunction = i;
+		_currentScriptContext->functions.push_back(new ScriptData);
+		_currentScript = _currentScriptContext->functions[_currentScriptFunction];
+
+		uint16 nameIndex = stream.readUint16();
+		stream.readUint16();
+		uint32 length = stream.readUint32();
+		uint32 startOffset = stream.readUint32();
+		uint16 argCount = stream.readUint16();
+		uint32 argOffset = stream.readUint32();
+		uint16 varCount = stream.readUint16();
+		uint32 varNamesOffset = stream.readUint32();
+		stream.readUint16();
+		stream.readUint16();
+		stream.readUint16();
+		stream.readUint16();
+		stream.readUint16();
+		stream.readUint16();
+		stream.readUint16();
+		stream.readUint16();
+		stream.readUint16();
+
+		if (startOffset < codeStoreOffset) {
+			warning("Function %d start offset is out of bounds!", i);
+			continue;
+		} else if (startOffset + length >= codeStoreOffset + codeStoreSize) {
+			warning("Function %d end offset is out of bounds", i);
+			continue;
+		}
+
+		uint16 pointer = startOffset - codeStoreOffset;
+		Common::Array<uint32> offsetList;
+		while (pointer < startOffset + length - codeStoreOffset) {
+			uint8 opcode = codeStore[pointer];
+
+			pointer += 1;
+			if (_lingoV4.contains(opcode)) {
+				offsetList.push_back(_currentScript->size());
+				g_lingo->code1(_lingoV4[opcode]->func);
+				size_t argc = strlen(_lingoV4[opcode]->proto);
+				for (uint c=0; c < argc; c++) {
+					switch (_lingoV4[opcode]->proto[c]) {
+						case 'b':
+							offsetList.push_back(_currentScript->size());
+							g_lingo->codeInt((int8)codeStore[pointer]);
+							pointer += 1;
+							break;
+						case 'w':
+							offsetList.push_back(_currentScript->size());
+							offsetList.push_back(_currentScript->size());
+							g_lingo->codeInt((int16)READ_UINT16(&codeStore[pointer]));
+							pointer += 2;
+							break;
+						default:
+							break;
+					}
+				}
+
+			} else {
+				// unimplemented instruction
+				if (opcode < 0x40) { // 1 byte instruction
+					offsetList.push_back(_currentScript->size());
+					g_lingo->code1(Lingo::c_unk);
```

Comment:
Should we perhaps print out the codes of the unimplemented instrucitons? Then we could see what is missing in the traces.

Reply:
c_unk/c_unk1/c_unk2 will print out warning traces when the missing instructions are executed, is it worthwhile adding more for the loader?

Reply:
Added